File: app/src/main/java/com/example/mynote/ui/pages/List/NotesListFragment.java
package com.example.mynote.ui.pages.List;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.PopupMenu;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.mynote.App;
import com.example.mynote.R;
import com.example.mynote.Impl.NotesDiffCallback;
import com.example.mynote.model.entities.NoteEntity;
import com.example.mynote.model.repos.NotesRepository;

@SuppressWarnings("FieldCanBeLocal")
public class NotesListFragment extends Fragment {
    private final NotesAdapter adapter = new NotesAdapter();
    private final Subscriber subscriber = this::onNoteSaved;

    private App app;
    private NotesRepository repository;

    private RecyclerView recyclerView;
    private Controller controller;

    public interface Controller {
        void openNoteEditScreen(NoteEntity item);

        void subscribe(Subscriber subscriber);

        void unsubscribe(Subscriber subscriber);
    }

    public interface Subscriber {
        void onNoteSaved(NoteEntity note);
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (context instanceof Controller) {
            controller = (Controller) context;
        } else {
            throw new IllegalStateException("Activity must implement NotesListFragment.Controller");
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_notes_list, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        controller.subscribe(subscriber);
        initRepository();
        initRecyclerView(view);
    }

    @Override
    public void onDestroy() {
        controller.unsubscribe(subscriber);
        controller = null;
        super.onDestroy();
    }

    private void initRepository() {
        app = (App) requireActivity().getApplication();
        repository = app.getNotesRepository();
    }

    private void initRecyclerView(View view) {
        recyclerView = view.findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(adapter);
        adapter.setData(repository.getNotes());
        adapter.setOnItemClickListener(new OnItemClickListener() {
            @Override
            public void onItemClick(NoteEntity item, int position) {
                controller.openNoteEditScreen(item);
            }

            @Override
            public void onItemLongClick(NoteEntity item, View view, int position) {
                showNotePopupMenu(item, view);
            }
        });
    }

    @SuppressLint("NonConstantResourceId")
    private void showNotePopupMenu(NoteEntity item, View view) {
        PopupMenu popupMenu = new PopupMenu(requireContext(), view, Gravity.END);
        popupMenu.inflate(R.menu.note_item_popup_menu);
        popupMenu.setOnMenuItemClickListener(menuItem -> {
            switch (menuItem.getItemId()) {
                case R.id.edit_popup_menu_item:
                    controller.openNoteEditScreen(item);
                    break;
                case R.id.delete_popup_menu_item:
                    deleteItem(item);
                    break;
            }
            return true;
        });
        popupMenu.show();
    }

    private void deleteItem(NoteEntity item) {
        if (repository.deleteNote(item.getUid())) {
            Toast.makeText(requireActivity(),
                    getString(R.string.successfully_deleted) + item.getTitle(),
                    Toast.LENGTH_SHORT).show();
            checkDiffRepo();
        } else {
            Toast.makeText(requireActivity(),
                    R.string.fail_to_delete,
                    Toast.LENGTH_SHORT).show();
        }
    }

    public void onNoteSaved(NoteEntity note) {
        if (note != null) {
            if (note.getUid() == null) {
                repository.createNote(note);
            } else {
                repository.updateNote(note.getUid(), note);
            }
            checkDiffRepo();
        }
    }

    public void checkDiffRepo() {
        NotesDiffCallback notesDiffCallback = new NotesDiffCallback(adapter.getData(), repository.getNotes());
        DiffUtil.DiffResult result = DiffUtil.calculateDiff(notesDiffCallback, true);
        adapter.setData(repository.getNotes());
        result.dispatchUpdatesTo(adapter);
    }
}

File: app/src/main/java/com/example/mynote/ui/pages/List/NotesListFragment.kt
package com.example.mynote.ui.pages.List;

import com.example.mynote.model.entities.NoteEntity
import com.example.mynote.App
import com.example.mynote.model.repos.NotesRepository
import android.os.Bundle
import androidx.recyclerview.widget.LinearLayoutManager
import android.annotation.SuppressLint
import android.content.Context
import android.view.*
import android.widget.PopupMenu
import com.example.mynote.R
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.example.mynote.Impl.NotesDiffCallback
import androidx.recyclerview.widget.DiffUtil
import com.example.mynote.databinding.FragmentNotesListBinding
import java.lang.IllegalStateException

class NotesListFragment : Fragment() {
    private val adapter = NotesAdapter()
    private val subscriber: Subscriber = object : Subscriber {
        override fun onNoteSaved(note: NoteEntity?) {
            if (note != null) {
                if (note.uid == null) {
                    repository!!.createNote(note)
                } else {
                    repository!!.updateNote(note.uid, note)
                }
                checkDiffRepo()
            }
        }
    }
    private var app: App? = null
    private var repository: NotesRepository? = null
    private var binding: FragmentNotesListBinding? = null
    private var controller: Controller? = null

    interface Controller {
        fun openNoteEditScreen(item: NoteEntity?)
        fun subscribe(subscriber: Subscriber?)
        fun unsubscribe(subscriber: Subscriber?)
    }
    interface Subscriber {
        fun onNoteSaved(note: NoteEntity?)
    }
    override fun onAttach(context: Context) {
        super.onAttach(context)
        controller = if (context is Controller) {
            context
        } else {
            throw IllegalStateException("Activity must implement NotesListFragment.Controller")
        }
    }
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentNotesListBinding.inflate(inflater, container, false)
        return binding!!.root
    }
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        controller!!.subscribe(subscriber)
        initRepository()
        initRecyclerView()
    }
    override fun onDestroyView() {
        binding = null
        super.onDestroyView()
    }
    override fun onDestroy() {
        controller!!.unsubscribe(subscriber)
        controller = null
        super.onDestroy()
    }
    private fun initRepository() {
        app = requireActivity().application as App
        repository = app!!.notesRepository
    }
    private fun initRecyclerView() {
        binding!!.recyclerView.layoutManager = LinearLayoutManager(context)
        binding!!.recyclerView.adapter = adapter
        adapter.data = repository!!.notes
        adapter.setOnItemClickListener(object : OnItemClickListener {
            override fun onItemClick(item: NoteEntity, position: Int) {
                controller!!.openNoteEditScreen(item)
            }
            override fun onItemLongClick(item: NoteEntity, view: View, position: Int) {
                showNotePopupMenu(item, view)
            }
        })
    }
    private fun showNotePopupMenu(item: NoteEntity, view: View) {
        val popupMenu = PopupMenu(requireContext(), view, Gravity.END)
        popupMenu.inflate(R.menu.note_item_popup_menu)
        popupMenu.setOnMenuItemClickListener { menuItem: MenuItem ->
            when (menuItem.itemId) {
                R.id.edit_popup_menu_item -> controller!!.openNoteEditScreen(item)
                R.id.delete_popup_menu_item -> deleteItem(item)
            }
            true
        }
        popupMenu.show()
    }
    private fun deleteItem(item: NoteEntity) {
        if (repository!!.deleteNote(item.uid)) {
            Toast.makeText(requireActivity(),
                    getString(R.string.successfully_deleted) + item.title,
                    Toast.LENGTH_SHORT).show()
            checkDiffRepo()
        } else {
            Toast.makeText(requireActivity(),
                    R.string.fail_to_delete,
                    Toast.LENGTH_SHORT).show()
        }
    }
    public fun onNoteSaved(note: NoteEntity?) {
        if (note != null) {
            if (note.uid == null) {
                repository!!.createNote(note)
            } else {
                repository!!.updateNote(note.uid, note)
            }
            checkDiffRepo()
        }
    }

    fun checkDiffRepo() {
        val notesDiffCallback = NotesDiffCallback(adapter.data, repository!!.notes)
        val result = DiffUtil.calculateDiff(notesDiffCallback, true)
        adapter.data = repository!!.notes
        result.dispatchUpdatesTo(adapter)
    }
}